File: scripts/populate-peak-content.ts
/**
 * Populer description, image_url og image_credit for alle fjelltopper.
 *
 * Kjør fra fjelltopper/-mappen:
 *   npx tsx scripts/populate-peak-content.ts             # full kjøring
 *   npx tsx scripts/populate-peak-content.ts --dry-run   # forhåndsvisning
 *   npx tsx scripts/populate-peak-content.ts --limit 5   # test med 5 topper
 *   npx tsx scripts/populate-peak-content.ts --skip-images # kun beskrivelser
 *
 * Beskrivelser leses fra scripts/peak-descriptions.json (generert av Claude).
 * Faller tilbake til mal-beskrivelse hvis ID ikke finnes i JSON.
 */

import fs from 'fs';
import path from 'path';

// Les inn forhåndsgenererte beskrivelser
const descriptionsPath = path.join(process.cwd(), 'scripts', 'peak-descriptions.json');
const PEAK_DESCRIPTIONS: Record<string, string> = JSON.parse(
  fs.readFileSync(descriptionsPath, 'utf-8')
);

const SUPABASE_URL = "https://enrbraxvxnytyfeuewqq.supabase.co";
const SERVICE_KEY =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImVucmJyYXh2eG55dHlmZXVld3FxIiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTc3ODg2ODY3MCwiZXhwIjoyMDk0NDQ0NjcwfQ.4vxGMgeIoBZEh9PoLcZdxq0-00jOMiIjyOUR7JHX-6s";

const HEADERS: Record<string, string> = {
  apikey: SERVICE_KEY,
  Authorization: `Bearer ${SERVICE_KEY}`,
  "Content-Type": "application/json",
  Prefer: "return=minimal",
};

// Filtrerer på filnavn (siste del av URL) — ikke hele pathen
// Holdes bevisst kort for å unngå å blokkere gyldige fjellbilder
const SKIP_KEYWORDS = [
  "map", "logo", "icon", "coat", "flag", "symbol",
  "locator", "topographic", "outline",
  "hytte", "hotel", "hytter", "lodge",
];

function isSkippable(url: string): boolean {
  // Hent kun filnavnet, ikke hele URL-pathen
  const filename = url.split("/").pop()?.toLowerCase() ?? "";
  return SKIP_KEYWORDS.some((kw) => filename.includes(kw));
}

// ---------------------------------------------------------------------------
// Args
// ---------------------------------------------------------------------------

const args = process.argv.slice(2);
const DRY_RUN     = args.includes("--dry-run");
const SKIP_IMAGES = args.includes("--skip-images");
const FORCE       = args.includes("--force"); // overskriver eksisterende innhold
const limitIdx    = args.indexOf("--limit");
const LIMIT       = limitIdx !== -1 ? parseInt(args[limitIdx + 1]) : null;

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

interface Peak {
  id: string;
  name: string;
  height: number;
  county: string | null;
  municipality: string | null;
  primary_factor: number | null;
  nearest_higher_peak: string | null;
  description: string | null;
  image_url: string | null;
}

// ---------------------------------------------------------------------------
// Beskrivelse — henter fra JSON, faller tilbake til mal
// ---------------------------------------------------------------------------

function generateDescription(peak: Peak): string {
  // Bruk forhåndsgenerert beskrivelse hvis tilgjengelig
  if (PEAK_DESCRIPTIONS[peak.id]) {
    return PEAK_DESCRIPTIONS[peak.id];
  }

  // Fallback: mal-beskrivelse
  const { height, county, municipality, primary_factor: pf, nearest_higher_peak: nhp } = peak;
  const sentences: string[] = [];

  if (municipality && county) {
    sentences.push(
      `Toppen ligger i ${municipality} kommune i ${county} og er ${height} meter over havet.`
    );
  } else if (county) {
    sentences.push(`Toppen ligger i ${county} og er ${height} meter over havet.`);
  } else {
    sentences.push(`Toppen er ${height} meter over havet.`);
  }

  if (pf != null) {
    if (pf >= 500) {
      sentences.push(
        `Med en primærfaktor på ${pf} meter er den et markant og selvstendig fjell i terrenget.`
      );
    } else if (pf >= 100) {
      sentences.push(
        `Primærfaktoren på ${pf} meter gjør den til en tydelig selvstendig topp.`
      );
    } else if (pf >= 30) {
      sentences.push(`Primærfaktoren er ${pf} meter.`);
    }
  }

  if (nhp) {
    sentences.push(`Nærmeste høyere topp er ${nhp}.`);
  }

  return sentences.join(" ");
}

// ---------------------------------------------------------------------------
// Wikimedia Commons
// ---------------------------------------------------------------------------

function stripHtml(text: string): string {
  return text.replace(/<[^>]+>/g, "").trim();
}

function sleep(ms: number) {
  return new Promise((r) => setTimeout(r, ms));
}

// Fjern vanlige norske retningssufikser for å bedre søketreff
// "Store Styggedalstinden Østtoppen" → "Store Styggedalstinden"
function baseName(name: string): string {
  return name
    .replace(/\s+(Østtoppen|Vesttoppen|Nordtoppen|Sørtoppen|Midttoppen|Fortoppen|Sørøsttoppen|Nordøsttoppen|Sørvesttoppen|Nordvesttoppen|N\d+|S\d+|V\d+|Ø\d+|SØ\d+|NV\d+|NØ\d+|SV\d+)$/i, "")
    .trim();
}

async function fetchWikimediaImage(
  peakName: string
): Promise<{ url: string; credit: string | null } | null> {
  const base = baseName(peakName);
  // Prøv norsk søk først — fungerer best for norske fjellnavn
  const queries = [
    base,
    `${base} fjell`,
    `${base} mountain`,
  ];
  // Ikke dupliser hvis basename == peakName
  if (base !== peakName) queries.push(peakName);

  for (const query of queries) {
    const params = new URLSearchParams({
      action: "query",
      generator: "search",
      gsrnamespace: "6",
      gsrsearch: query,
      gsrlimit: "20",          // økt fra 8 → 20
      prop: "imageinfo",
      iiprop: "url|extmetadata|mime",
      iiurlwidth: "800",
      format: "json",
      origin: "*",
    });

    try {
      const resp = await fetch(
        `https://commons.wikimedia.org/w/api.php?${params}`
      );
      const text = await resp.text();
      let data: any;
      try {
        data = JSON.parse(text);
      } catch {
        await sleep(2000);
        continue;
      }
      const pages: Record<string, any> = data?.query?.pages ?? {};

      for (const page of Object.values(pages)) {
        const ii = page?.imageinfo?.[0];
        if (!ii) continue;

        const mime: string = ii.mime ?? "";
        const url: string  = ii.thumburl ?? ii.url ?? "";

        if (!mime.startsWith("image/jpeg") && !mime.startsWith("image/png")) continue;
        if (isSkippable(url)) continue;

        const meta        = ii.extmetadata ?? {};
        const artist      = stripHtml(meta?.Artist?.value ?? "");
        const licenseName = meta?.LicenseShortName?.value ?? "";
        const credit      = [artist, licenseName].filter(Boolean).join(" / ") || null;

        return { url, credit };
      }
    } catch (e) {
      console.error(`    ⚠ Wikimedia-feil for '${query}': ${e}`);
    }

    await sleep(200);
  }

  return null;
}

// ---------------------------------------------------------------------------
// Supabase
// ---------------------------------------------------------------------------

async function fetchPeaks(): Promise<Peak[]> {
  const params = new URLSearchParams({
    select:
      "id,name,height,county,municipality,primary_factor,nearest_higher_peak,description,image_url",
    height: "gte.2000",
    order:  "height.desc",
    limit:  "500",
  });

  const resp = await fetch(`${SUPABASE_URL}/rest/v1/peaks?${params}`, {
    headers: HEADERS,
  });

  if (!resp.ok) {
    throw new Error(`Supabase feil: ${resp.status} ${await resp.text()}`);
  }
  return resp.json();
}

async function updatePeak(
  id: string,
  payload: Record<string, unknown>
): Promise<boolean> {
  const resp = await fetch(`${SUPABASE_URL}/rest/v1/peaks?id=eq.${id}`, {
    method:  "PATCH",
    headers: HEADERS,
    body:    JSON.stringify(payload),
  });
  return resp.status === 200 || resp.status === 204;
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function main() {
  if (DRY_RUN)  console.log("🔍 DRY-RUN — ingenting skrives til Supabase\n");

  console.log("Henter topper fra Supabase...");
  const peaks = await fetchPeaks();
  console.log(`Fant ${peaks.length} topper totalt`);

  let toProcess = FORCE
    ? peaks
    : peaks.filter((p) => !p.description || !p.image_url);
  console.log(`Skal behandle: ${toProcess.length} topper${FORCE ? " (--force, inkl. eksisterende)" : " (mangler innhold)"}\n`);

  if (LIMIT) {
    toProcess = toProcess.slice(0, LIMIT);
    console.log(`Begrenset til ${toProcess.length} topper (--limit)\n`);
  }

  let okDesc = 0, okImg = 0, errors = 0;

  for (let i = 0; i < toProcess.length; i++) {
    const peak = toProcess[i];
    console.log(`[${i + 1}/${toProcess.length}] ${peak.name} (${peak.height} m)`);

    const payload: Record<string, unknown> = {};

    // Beskrivelse
    if (!peak.description || FORCE) {
      const desc = generateDescription(peak);
      payload.description = desc;
      okDesc++;
      console.log(`    📝 ${desc.slice(0, 90)}`);
    }

    // Bilde
    if ((!peak.image_url || FORCE) && !SKIP_IMAGES) {
      const img = await fetchWikimediaImage(peak.name);
      if (img) {
        payload.image_url    = img.url;
        payload.image_credit = img.credit;
        okImg++;
        console.log(`    🖼  ${img.url.slice(0, 70)}`);
        if (img.credit) console.log(`    ©  ${img.credit}`);
      } else {
        console.log(`    – Ingen Wikimedia-bilde funnet`);
      }
      await sleep(300);
    }

    // Skriv til Supabase
    if (Object.keys(payload).length > 0) {
      if (!DRY_RUN) {
        const ok = await updatePeak(peak.id, payload);
        if (!ok) {
          console.error(`    ❌ Feil ved oppdatering av ${peak.name}`);
          errors++;
        }
      } else {
        console.log(`    [DRY-RUN] ville skrevet: ${Object.keys(payload).join(", ")}`);
      }
    }
  }

  console.log(`\n✅ Ferdig!`);
  console.log(`   Beskrivelser skrevet: ${okDesc}`);
  console.log(`   Bilder hentet:        ${okImg}`);
  if (errors) console.log(`   Feil:                 ${errors}`);
  if (DRY_RUN) console.log(`\n   (DRY-RUN — ingenting ble skrevet til Supabase)`);
}

main().catch((e) => {
  console.error("Uventet feil:", e);
  process.exit(1);
});

File: scripts/populate-peak-content.py
"""
Populer description, image_url og image_credit for alle fjelltopper.

Kjør fra fjelltopper/-mappen:
    python scripts/populate-peak-content.py           # full kjøring
    python scripts/populate-peak-content.py --dry-run # forhåndsvisning uten skriving
    python scripts/populate-peak-content.py --limit 5 # test med 5 topper

Krever: pip install requests
"""

import requests
import json
import time
import re
import sys
import argparse

SUPABASE_URL = "https://enrbraxvxnytyfeuewqq.supabase.co"
SERVICE_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImVucmJyYXh2eG55dHlmZXVld3FxIiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTc3ODg2ODY3MCwiZXhwIjoyMDk0NDQ0NjcwfQ.4vxGMgeIoBZEh9PoLcZdxq0-00jOMiIjyOUR7JHX-6s"

HEADERS = {
    "apikey": SERVICE_KEY,
    "Authorization": f"Bearer {SERVICE_KEY}",
    "Content-Type": "application/json",
    "Prefer": "return=minimal"
}


# ---------------------------------------------------------------------------
# Beskrivelsesmal
# ---------------------------------------------------------------------------

def generate_description(peak: dict) -> str:
    name        = peak["name"]
    height      = peak["height"]
    municipality = peak.get("municipality") or ""
    county      = peak.get("county") or ""
    pf          = peak.get("primary_factor")
    nhp         = peak.get("nearest_higher_peak")

    sentences = []

    # Setning 1 — plassering og høyde
    if municipality and county:
        sentences.append(f"Toppen ligger i {municipality} kommune i {county} og er {height} meter over havet.")
    elif county:
        sentences.append(f"Toppen ligger i {county} og er {height} meter over havet.")
    else:
        sentences.append(f"Toppen er {height} meter over havet.")

    # Setning 2 — primærfaktor
    if pf is not None:
        if pf >= 500:
            sentences.append(
                f"Med en primærfaktor på {pf} meter er den et markant og selvstendig fjell i terrenget."
            )
        elif pf >= 100:
            sentences.append(
                f"Primærfaktoren på {pf} meter gjør den til en tydelig selvstendig topp."
            )
        elif pf >= 30:
            sentences.append(
                f"Primærfaktoren er {pf} meter."
            )

    # Setning 3 — nærmeste høyere topp
    if nhp:
        sentences.append(f"Nærmeste høyere topp er {nhp}.")

    return " ".join(sentences)


# ---------------------------------------------------------------------------
# Wikimedia Commons bildesøk
# ---------------------------------------------------------------------------

SKIP_KEYWORDS = ["map", "logo", "icon", "coat", "flag", "symbol",
                 "locator", "kart", "location", "outline", "topographic"]

def strip_html(text: str) -> str:
    return re.sub(r"<[^>]+>", "", text).strip()

def fetch_wikimedia_image(peak_name: str) -> tuple[str | None, str | None]:
    """Returnerer (image_url, image_credit) eller (None, None)."""
    queries = [
        f"{peak_name} mountain norway",
        f"{peak_name} fjell norge",
        peak_name,
    ]

    for query in queries:
        try:
            resp = requests.get(
                "https://commons.wikimedia.org/w/api.php",
                params={
                    "action":       "query",
                    "generator":    "search",
                    "gsrnamespace": "6",
                    "gsrsearch":    query,
                    "gsrlimit":     "8",
                    "prop":         "imageinfo",
                    "iiprop":       "url|extmetadata|mime",
                    "iiurlwidth":   "800",
                    "format":       "json",
                    "origin":       "*",
                },
                timeout=10
            )
            data = resp.json()
        except Exception as e:
            print(f"    ⚠ Wikimedia-feil for '{query}': {e}")
            continue

        pages = data.get("query", {}).get("pages", {})
        for page in pages.values():
            ii = page.get("imageinfo", [{}])[0]
            mime = ii.get("mime", "")
            url  = ii.get("thumburl") or ii.get("url", "")

            # Filtrer ut ikke-bilder og uønskede filer
            if not mime.startswith(("image/jpeg", "image/png")):
                continue
            url_lower = url.lower()
            if any(kw in url_lower for kw in SKIP_KEYWORDS):
                continue

            # Bygg krediteringsstreng
            meta    = ii.get("extmetadata", {})
            artist  = strip_html(meta.get("Artist", {}).get("value", ""))
            license_name = meta.get("LicenseShortName", {}).get("value", "")
            credit  = " / ".join(filter(None, [artist, license_name])) or None

            return url, credit

        time.sleep(0.2)

    return None, None


# ---------------------------------------------------------------------------
# Supabase-operasjoner
# ---------------------------------------------------------------------------

def fetch_peaks() -> list[dict]:
    resp = requests.get(
        f"{SUPABASE_URL}/rest/v1/peaks",
        headers=HEADERS,
        params={
            "select": "id,name,height,county,municipality,primary_factor,nearest_higher_peak,description,image_url",
            "height": "gte.2000",
            "order":  "height.desc",
            "limit":  "500",
        },
        timeout=15
    )
    resp.raise_for_status()
    return resp.json()


def update_peak(peak_id: str, payload: dict) -> bool:
    resp = requests.patch(
        f"{SUPABASE_URL}/rest/v1/peaks",
        headers=HEADERS,
        params={"id": f"eq.{peak_id}"},
        json=payload,
        timeout=15
    )
    return resp.status_code in (200, 204)


# ---------------------------------------------------------------------------
# Hovedløkke
# ---------------------------------------------------------------------------

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--dry-run", action="store_true",
                        help="Vis hva som ville blitt skrevet uten å oppdatere Supabase")
    parser.add_argument("--limit", type=int, default=None,
                        help="Begrens antall topper som behandles")
    parser.add_argument("--skip-images", action="store_true",
                        help="Hopp over Wikimedia-bildesøk")
    args = parser.parse_args()

    print("Henter topper fra Supabase...")
    peaks = fetch_peaks()
    print(f"Fant {len(peaks)} topper totalt")

    # Filtrer kun de som mangler innhold
    to_process = [p for p in peaks if not p.get("description") or not p.get("image_url")]
    print(f"Mangler innhold: {len(to_process)} topper\n")

    if args.limit:
        to_process = to_process[:args.limit]
        print(f"Begrenset til {len(to_process)} topper (--limit)\n")

    ok_desc = ok_img = errors = 0

    for i, peak in enumerate(to_process):
        name = peak["name"]
        print(f"[{i+1}/{len(to_process)}] {name} ({peak['height']} m)")

        payload = {}

        # Beskrivelse
        if not peak.get("description"):
            desc = generate_description(peak)
            payload["description"] = desc
            ok_desc += 1
            print(f"    📝 {desc[:80]}...")

        # Bilde
        if not peak.get("image_url") and not args.skip_images:
            img_url, credit = fetch_wikimedia_image(name)
            if img_url:
                payload["image_url"] = img_url
                payload["image_credit"] = credit
                ok_img += 1
                print(f"    🖼  {img_url[:70]}...")
            else:
                print(f"    – Ingen Wikimedia-bilde funnet")
            time.sleep(0.3)

        # Skriv til Supabase
        if payload and not args.dry_run:
            success = update_peak(peak["id"], payload)
            if not success:
                print(f"    ❌ Feil ved oppdatering av {name}")
                errors += 1
        elif args.dry_run and payload:
            print(f"    [DRY-RUN] ville skrevet: {list(payload.keys())}")

    print(f"\n✅ Ferdig!")
    print(f"   Beskrivelser skrevet: {ok_desc}")
    print(f"   Bilder hentet:        {ok_img}")
    print(f"   Feil:                 {errors}")
    if args.dry_run:
        print("   (DRY-RUN — ingenting ble faktisk skrevet til Supabase)")


if __name__ == "__main__":
    main()
